src/main/java/net/starlark/java/syntax/Resolver.java

```diff
@@ -982,6 +982,9 @@ private void checkLoadAfterStatement(List<Statement> statements) {
    * StarlarkFile#errors}.
    */
   public static void resolveFile(StarlarkFile file, Module module) {
+    Preconditions.checkState(!file.resolveStarted, "StarlarkFile can be resolved only once");
```

Comment:
Hmm, why not just `Preconditions.checkState(file.getResolvedFunction() == null, ...)`?

What scenario are we trying to protect against? If it's duplicate sequential calls to resolveFile - a check for whether file.getResolvedFunction() is null ought to be enough. If it's interleaved calls to resolveFile from two threads - you would need real locking.

Reply:
> why not just

Resolve may fail before it sets `resolvedFunction`. It will poison `StarlarkFile` state, and consequent call to resolve will be harder to understand.

For example, client code might want to do something like:

```
if (file.getResolvedFunction() == null) {
  resolve
}
```

> What scenario are we trying to protect against?

Make API easier to use: if we accidentally call resolve twice for some reason, better crash explicitly.

At the time I was struggling to understand how API works, so I decided to add some validation.

Anyway, this PR is not a big deal. I can abandon it.